File: app/channels/rooms_channel.rb
class RoomsChannel < ApplicationCable::Channel
  def subscribed
    current_user.rooms.each do |room|
      stream_from "rooms:#{room.id}"
    end
  end

  def unsubscribed
    stop_all_streams
  end

  def send_message(data)
    @room = Room.find(data['room_id'])
    message = @room.messages.create(body: data['body'], user: current_user)

    MessageRelayJob.perform_later(message, current_user)
  end

  def send_last_read(data)
    room_user = current_user.room_users.find_by(room_id: data['room_id'])
    room_user.update(last_read_at: Time.zone.now)
  end
end

File: app/javascript/channels/rooms_channel.js
import consumer from "./consumer"

function getChatContainer(data) {
  return document.querySelectorAll(`[data-chat-room="${data.room_id}"]`)
}

function getMessages(chat) {
  return chat[0].querySelectorAll(`[data-behavior="messages"]`)
}

function getRoomLink(data) {
  return document.querySelectorAll(`[data-behavior="room-link"][data-room-id="${data.room_id}"]`)
}

export default consumer.subscriptions.create("RoomsChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    const chat = getChatContainer(data)
    if (chat.length > 0) {
      const messages = getMessages(chat)
      if (messages.length > 0) {
        if (document.hidden) {
          if (document.getElementById("unread") === null) {
            messages[0].insertAdjacentHTML("beforeend", data.divider)
          }

          if (Notification.permission === "granted") {
            new Notification(data.user.email, { body: data.body })
          }
        } else {
          this.sendLastRead({ room: data.room_id })
        }

        messages[0].insertAdjacentHTML("beforeend", data.message)
        messages[0].scrollTop = messages[0].scrollHeight
      } else {
        const roomLink = getRoomLink(data)
        if (roomLink.length > 0) {
          roomLink[0].classList.add("font-semibold")
        }
      }
    }
  },

  sendMessage(data) {
    this.perform("send_message", {
      room_id: data.room,
      body: data.body,
    })
  },

  sendLastRead(data) {
    this.perform("send_last_read", { room_id: data.room })
  },
})
